### ISSUE TITLE
useLayoutEffect in ssr

### ISSUE THREAD

**Issue by dimensi** [facebook/react#14927]
Hi, I do not understand the situation with this hook a bit. I use this hook to perform the animation synchronously with the state update, if I use useEffect, then I have jumps in the animation, because the animation library does not have time to start. Also, the documentation states that useLayoutEffect runs on the same phase as componentDidMount (that is, on the client side), and here my server issues complaints to me about my code. Why is that?

https://codesandbox.io/s/oo47nj9mk9

_Originally posted by @dimensi in https://github.com/facebook/react/pull/14596#issuecomment-466023638_

**Comment by sebmarkbage** [comment]
It’s there to force you to think about whether you truly need it to be useLayoutEffect (uncommon) or if you’re ok with it being useEffect (common).

useLayoutEffect exists to give you strong guarantees about having the ability to adjust layout last minute without painting between. However we cannot guarantee that if you’re server rendering this component. It has to be resilient to work without this guarantee.

**Comment by thysultan** [comment]
Related thread on some of the effects of this in practice: [Twitter Link (https://twitter.com/probablyup/status/1097507647371296768).

**Comment by sebmarkbage** [comment]
I can imagine some theoretical cases where it would be legit to want to useLayoutEffect only after some state has switched.

However in practice I haven’t seen this yet. All cases I’ve seen so far have been slightly broken when you SSR if you think about it.

It would be good to show actual examples if you have them.

**Comment by dimensi** [comment]
@sebmarkbage 
https://codesandbox.io/s/7y96862n2q
try this.

**Comment by atomiks** [comment]
I'm integrating a 3rd party lib that mutates DOM elements in the tree, and it needs to happen before first paint to prevent a small "jitter". This warning occurs during SSR, but as mentioned above, I don't see why if it's equivalent to the cDM/cDU lifecycles that never did.

**Comment by atomiks** [comment]
Maybe this could work or would it break the rules of hooks?

```js
const effect = isBrowser ? useLayoutEffect : useEffect

function Comp() {
  effect(() => {
    // ...
  })
}
```

**Comment by sebmarkbage** [comment]
@atomiks How does that component work with server rendering? Isn’t there a small jitter?

**Comment by atomiks** [comment]
@sebmarkbage it's a tooltip library that creates a separate element that only exists after mounting. The element it's attached to gets rendered normally (server or client), but the tooltip doesn't appear until after hydration.

<details>
<summary>Details</summary>
The content is rendered in a portal. When the content gets updated by React, the `.set()` method should be be called synchronously for the position to be updated. Because it uses position: absolute and doesn't exist in the normal flow of the document, changes to the size of the tooltip causes it to be repositioned in the wrong place, so the `translate` transform needs to be updated immediately before painting, or there will be a jitter.
 
A ResizeObserver could also work, but not enough support yet.

Demo with `useEffect`

<img src="https://i.gyazo.com/b8b167cb105c50e4595314aa2be06438.gif" width="400" alt="jitters" />

- Happens once on first load, twice on second load
- When it reaches the boundary, it doesn't happen because the position doesn't need to be updated


No jitters with `useLayoutEffect`
</details>

**Comment by dimensi** [comment]
Any answers?

**Comment by atomiks** [comment]
@dimensi I am using the trick in https://github.com/facebook/react/issues/14927#issuecomment-466815232 and it seems to be working fine for my component library. I had to release a patch just to get around it.

I don't see why there's even a warning at all, what's the point? We need to use hacks to get around it anyway because it _is_ required in many legitimate cases, but still want to use it in SSR without problems (no-op behavior is fine anyway)...

**Comment by gaearon** [comment]
>what's the point

The point of the warning is to warn you that your component will behave weirdly before hydration. You may disagree or ignore the recommendation in specific cases if you want but it's pointing out a legitimate problem.

**Comment by gaearon** [comment]
If something only exists after hydration then the canonical solution to it is to render a fallback view first. And then flip a flag inside `useEffect` that would show and activate your plugin.

```js
function useIsMounted() {
  let [mounted, setMounted] = useState(false);
  useEffect(() => {
    setMounted(true);
  }, []);
  return mounted;
}

function Foo() {
  let ref = useRef();
  let isMounted = useIsMounted();
  useEffect(() => {
    if (isMounted) {
      $.myJqueryPlugin(ref.current);
    }
  }, [isMounted]);
  if (!isMounted) {
    return <FallbackView />;
  }
  return <div ref={ref} />;
}
```

**Comment by atomiks** [comment]
It's not a problem in this case though, as mentioned above. And I still need `useLayoutEffect` on updates, as mentioned above also. There's no way to get around it..

**Comment by gaearon** [comment]
>I think warnings that can't be turned off when there are use cases for it should not exist

I generally agree with that. That’s why the issue is still open. If there are legit use cases we’ll want to either adjust the warning or write documentation for it. Currently there are higher priority issues so we’re looking into this first. Since you can work around. We’ll get back to this and fix it after those.

**Comment by atomiks** [comment]
One thing that's a bit worrying with `useEffect` is that people are using it by default for a lot of layout-related things, unknowingly. Even things that don't seem layout-related, like `window.scrollTo`. I noticed in different threads where people are unexpectedly creating jittery UIs as a side effect now. And the worst part is it can be reproduced <50% of the time based on when the browser decides to paint when rendering, so it can be harder to catch. 

I think the differences between the hooks should be highlighted better (when to use them, etc.)

**Comment by gaearon** [comment]
The thing with jitter is that once you see it, you can fix it. With a layout effect. Or by calculating the correct initial state early (if you jitter due to a setState). It’s more annoying to first experience it — but **it’s intentional that sometimes you’d see it and that would tell you “okay, here I need a layout effect instead”.** This is feature working as designed. 

However the majority of code people put into DidMount/DidUpdate actually doesn’t need to be sync. So it’s a worse perf default which becomes death by a thousand cuts in a larger trees. 

In the past all code was sync. This has bad consequences for perf. Now it’s usually async, and when it causes jitter you just opt back into sync for that specific case. I think this makes sense. The naming is intentionally shorter for useEffect because we *want* you to try it first. And only replace it it actually causes a problem. (For most application code it doesn’t.)

I agree it can be annoying if you see a warning which can’t be fixed. Collecting more use cases that seem legit here would help.

For example I’m not sure the scroll one is legit if you use SSR. Scroll *does* represent a “layout effect” to me because you want it to happen *together* with layout instead or flickering. But what about SSR? You don’t want to start interacting with a page, scroll it, and then hydration scrolls it back. This is bad UX. So maybe your initial scroll logic should be outside of React components altogether, and could be inline in HTML. Then you’d be sure not to “scroll to top” too late. What do you think? More detailed use cases like this would help.

**Comment by eps1lon** [comment]
This should probably be mentioned in the docs: https://reactjs.org/docs/hooks-reference.html#uselayouteffect

That section makes it pretty clear to me that I should use `useLayoutEffect` if I migrate from class components. It doesn't mention the warning for server side rendering though.

Happy to submit a docs PR if I'm not the only one that finds the current documentation confusing.

**Comment by gaearon** [comment]
Sure, to be honest I want to rewrite that whole page because a lot of people find different parts about it confusing.

**Comment by gaearon** [comment]
I tweaked the docs to explain why this warning exists: https://reactjs.org/docs/hooks-reference.html#uselayouteffect

<img width="937" alt="docs screenshot" src="https://user-images.githubusercontent.com/810438/54627453-618b8380-4a6b-11e9-8b49-57265a687ed2.png">

**Comment by gaearon** [comment]
@dimensi 

Regarding your example in https://github.com/facebook/react/issues/14927#issuecomment-466703167. You didn't provide any details at all. What are the steps I need to do? What is the expected behavior? What is the actual behavior? Can you make a screenshot of the issue?

It's very hard to help when you provide a couple hundred lines with no explanation of what problem you're running into.

**Comment by markerikson** [comment]
It's starting to look like React-Redux's v7 implementation of `connect` may need to call `useLayoutEffect()` internally to avoid a timing issue.

The repro on this is somewhere between "complex" and "convoluted", but it involves a mixture of a sync `setState()` in a parent and a dispatched Redux action combined with `unstable_batchedUpdates()`, all in the same tick.  


I'm not sure I can even come up with a good TL;DR: for this one.  @alexreardon has repro sandboxes linked in https://github.com/reduxjs/react-redux/issues/1177#issuecomment-474638250 , and I wrote up a step-by-step description of what's going on in https://github.com/reduxjs/react-redux/issues/1177#issuecomment-474671590 .

`useLayoutEffect()` seems to resolve the issue, because it guarantees that a ref I'm using for coordination will be written to before any other logic (like a Redux store subscription) would possibly execute.

But, given that React-Redux is widely used for SSR, I don't want all our users to be seeing this warning printed all the time.

**Comment by alexreardon** [comment]
For timing reasons, `react-beautiful-dnd` leans heavily on `useLayoutEffect`. We need to tightly control user input event flows. My understanding would be that none of these functions would run in an SSR environment. It is a surprise to me that this would log a warning.

`react-beautiful-dnd` supports SSR and having these warnings would be lame for consumers

**Comment by alexreardon** [comment]
- [example 1: used to immediately stop capturing window input events when a drag is aborted](https://github.com/atlassian/react-beautiful-dnd/blob/hooks/src/view/use-drag-handle/use-drag-handle.js#L160)
- [example 2: used to immediately stop capturing window events when unmounting](https://github.com/atlassian/react-beautiful-dnd/blob/hooks/src/view/use-drag-handle/sensor/use-mouse-sensor.js#L372)
- [example 3: component registration. Immediately notify a marshal of mount changes](https://github.com/atlassian/react-beautiful-dnd/blob/hooks/src/view/use-draggable-dimension-publisher/use-draggable-dimension-publisher.js#L52)

**Comment by sebmarkbage** [comment]
For the `react-beautiful-dnd` case, if that was a leaf component, I'd say only conditionally render the component after initial render and render a fallback first since nothing will actually work until it's hydrated anyway. In fact, this is a good example for using a lazy component to load the richer functionality in lazily instead of paying for that during initial render. Since apparently it's fine to have a gap where this functionality doesn't work (which happens during SSR).

However you probably don't want to swap out the tree when this happens and you can't really swap out the component because it's a custom Hook.

This sounds like a good use case for "progressively enhancing hooks" which is a thing we've been thinking about. It's basically a way to let you load new Hooks into a component while it's already running. These new Hooks can then useLayoutEffect because they're loaded late and never during SSR. This guarantees that even in the client case, this speeds up initial rendering and still forces the component to deal with the gap between initial render and progressive enhancement.

**Comment by sebmarkbage** [comment]
For the Redux case I believe this is related to https://github.com/facebook/react/pull/15122#issuecomment-473591116

useEffect is basically just a concurrent mode-light and if you can't deal with the gap between rendering completing and commit phase, you're probably going to have even bigger problems when each render can also yield and gets dispatches injected in the middle.

**Comment by dimensi** [comment]
@gaearon 
The essence of my example is that I have 2 modes of animation work there: through useEffect and via useLayoutEffect.
If you select the useEffect mode (which is the default) and switch between slides, you can see how the text "jumps" and because animejs does not have time to apply all the styles when rendering the component. If you switch to useLayoutEffect, then there are no such problems, the animation works smoothly and the text does not "jump".
With this example, I want to show how important I useLayoutEffect for animation and that it does not affect the first render of the component.
And no, I don't want to use any dirty tricks to bypass the warning that creates reacts when rendering my component.
Because it is important for me that the content that is in the components was available immediately, and not after the initialization of react on the client. If I use tricks with stubs or conditions, then I will generally lose the sense of using ssr.

**Comment by salvoravida** [comment]
@gaearon @sebmarkbage 

as https://github.com/reduxjs/react-redux/issues/1177#issuecomment-474784122

i suggest IMHO, that you explain in hooks docs 
that ref.current updates should not be updates inside useEffect (as it may be deferred)

ref.current should be update as soon as there is a new value, like "last props received"
and of course it is fine to save it in the render phase.

That's all
Regards

**Comment by gaearon** [comment]
@dimensi Seems like your case is also the use case for “progressively enhancing” Hooks. (We don’t yet have an official solution to it but it’s something we’ve wanted to add.)

**Comment by markerikson** [comment]
@sebmarkbage : yeah, I think I'm inclined to agree on both points.

This _was_ one of the primary reasons why I switched us over to state propagation via context in React-Redux v6, but.... well, perf and hook updates, as we've talked about before :(  I'd happily stick with our current v6 approach if context worked the way we needed it to, but it doesn't sound like that's going to happen at this point.

v7 with direct subscriptions and use of `unstable_batchedUpdates` is working right now, but I agree that Bad Things (TM) are likely to happen with Concurrent Mode.  We'll deal with that whenever CM is finally ready.  I gotta come up with something that solves the issues our users are facing right now.

**Comment by brandonburkett** [comment]
My SSR logs :(

For attaching event listeners, per the docs, it seems I should use `useLayoutEffect`.  This use case is for main navigation (I'm converting from a class component) and auto-closing the menu if the user clicks outside the navigation menu component.  For simplicity, I wouldn't want to do something odd just to get around the warning message on SSR (as I didn't need too with the class based component).

**EX**
```
  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }
```

**Comment by WojciechJerczynski** [comment]
It is kinda annoying that warning as it spams the hell out of my tests even thought they are passing most of them come from react and react-redux connect method. Anyone came across it? Any way I can get rid of them?

**Comment by markerikson** [comment]
@VicJer : v7 actually _tries_ to fall back to `useEffect` in an SSR scenario specifically to avoid this.  Are you still seeing those warnings in a test environment?

**Comment by VicJer** [comment]
> @VicJer : v7 actually _tries_ to fall back to `useEffect` in an SSR scenario specifically to avoid this. Are you still seeing those warnings in a test environment?
@markerikson I am on 7.0.1 yeah I can still see it.

**Comment by atomiks** [comment]
@VicJer I'm guessing you're adding `global.window = someValue` to the Node environment then. I ended up adding `&& typeof document !== 'undefined'` for more safety, but if you're also declaring a global `document` then that [also won't help](https://github.com/FezVrasta/popper.js/pull/768). Usually `document` gets namespaced under `global.window.document` if browser env is being simulated though so it should work most of the time.

**Comment by VicJer** [comment]
I can't see it being added anywhere if I am honest. But I think I narrowed it down and it seems to me like it's Enzyme.

**Comment by eps1lon** [comment]
What about 

```js
React.useLayoutEffect(() => {
  if (autoFocus) {
    listItemRef.current.focus();
  }
}, [autoFocus]);
```
? `focus()` could change visuals of `listItemRef.current`. Valid use case to trick React into thinking that this is a `useEffect` on the server or not?

**Comment by alexreardon** [comment]
@sebmarkbage right now `react-beautiful-dnd` cannot move to a *"progressively enhancing hooks"* pattern as a `Droppable` currently sets up `context` for a `Draggable` - something a hook cannot do

**Comment by menberg** [comment]
I'm using [Overmind](overmindjs.org) for state management. While this library is made for the browser, I also can use it in a node environment. As the library utilises useLayoutEffect for some client side optimisations, I get the warning: **useLayoutEffect does nothing on the server**. Can I mute the warning somehow as my Terminal gets spammed?

**Comment by mandarzope** [comment]
When ever I pass initialState ={product:{....}} to createStore I get following error on server side render. If I make  initialState = {} error goes away.

Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://fb.me/react-uselayouteffect-ssr for common fixes.
    in ConnectFunction
    in ConnectFunction
    in div
    in Product
    in Context.Provider
    in ConnectFunction
    in ConnectFunction
    in Context.Provider
    in Context.Consumer
    in Route
    in Fragment
    in App
    in Context.Provider
    in Router
    in StaticRouter
    in Context.Provider
    in Provider

It is because connect()() function is using useLayoutEffect internally ?

**Comment by cjolowicz** [comment]
@mandarzope react-redux uses `useLayoutEffect` when `window` is defined:

https://github.com/reduxjs/react-redux/blob/8605088606403f6909597405c67db91d907db86e/src/components/connectAdvanced.js#L35-L41

By default, Jest defines `window` globally. This can lead to the `useLayoutEffect does nothing on the server` warning when running the test suite.

You can change this behaviour by selecting the `node` [test environment](https://jestjs.io/docs/en/configuration#testenvironment-string) instead of the default `jsdom` environment. Try adding a `@jest-environment` docblock to the very top of your test file:

```js
/**
 * @jest-environment jsdom
 */
```

Or, to select the `node` environment globally, use this in your `package.json`:

```json
{
  "name": "my-project",
  "jest": {
    "testEnvironment": "node"
  }
}```

The warning about `useLayoutEffect` should disappear because `window` is no longer defined during test execution.

**Comment by cezarsmpio** [comment]
I couldn't find any solution to use `useEffect` instead of `useLayoutEffect` in Safari when injecting CSS code on the fly. 

That's my hook, it works fine for all browsers when using `useLayoutEffect`:
```js
const cachedStyles = [];

function useInjectStyle(rule) {
  useLayoutEffect(function() {
    if (cachedStyles.indexOf(rule) >= 0) return;

    cachedStyles.push(rule);

    const styleElement = document.createElement('style');
    styleElement.appendChild(document.createTextNode(''));
    styleElement.setAttribute('react-fake-content', '');

    document.head.appendChild(styleElement);

    styleElement.sheet.insertRule(rule, styleElement.sheet.cssRules.length);
  }, []);
}
```
If I try to use CSS animations, it doesn't work in Safari using `useEffect` but it does work if I use `useLayoutEffect`.

```js
const keyframesRule = `@keyframes awesomeAnimationName { from { background-position: 0 center; } to { background-position: -200% center; } }`;
useInjectStyle(keyframesRule);
```

It simply doesn't add the animation only in Safari, all other browsers are fine with `useEffect` apparently.

**Comment by eps1lon** [comment]
One of the more popular [patterns to avoid invalidating `useCallback` too often](https://github.com/facebook/react/issues/14099#issuecomment-440013892) will also trigger this warning

**Comment by uberska** [comment]
We ran into this warning when we were using jest and enzyme to test a component with a child connected component. We had tests for the component that called enzyme's `mount` and `render` in the same file. `mount` requires a DOM, so we used jest's jsdom testEnvironment. When we upgraded React, the test that calls `render` started showing the warning about `useLayoutEffect`. react-redux was choosing `useLayoutEffect` in the child connected component because it's in a browser-like environment due to jest's jsdom testEnvironment. Since enzyme's `render` uses `ReactDOMServer`, we got the useLayoutEffect warning.

The solution is straightforward once you understand what's going on. I split the tests into two files. One file has the jsdom testEnvironment pragma and contains all the calls to `mount`. The other file has the node testEnvironment pragma and contains all the calls to `render`. If there are any ideas for keeping the tests in one file, let me know. I kind of like having a one-to-one mapping from component to component test file, but it's obviously not required nor critical.

**Comment by alexlee-dev** [comment]
> We ran into this warning when we were using jest and enzyme to test a component with a child connected component. We had tests for the component that called enzyme's `mount` and `render` in the same file. `mount` requires a DOM, so we used jest's jsdom testEnvironment. When we upgraded React, the test that calls `render` started showing the warning about `useLayoutEffect`. react-redux was choosing `useLayoutEffect` in the child connected component because it's in a browser-like environment due to jest's jsdom testEnvironment. Since enzyme's `render` uses `ReactDOMServer`, we got the useLayoutEffect warning.
> 
> The solution is straightforward once you understand what's going on. I split the tests into two files. One file has the jsdom testEnvironment pragma and contains all the calls to `mount`. The other file has the node testEnvironment pragma and contains all the calls to `render`. If there are any ideas for keeping the tests in one file, let me know. I kind of like having a one-to-one mapping from component to component test file, but it's obviously not required nor critical. Thanks!

Can you reference the 2 test files here for an example of how you solved this?

**Comment by steida** [comment]
> I can imagine some theoretical cases where it would be legit to want to useLayoutEffect only after some state has switched.

I believe such a case is anything related to the selection aka client only browser state.
A component could need to get and selection asap and it's actually a side effect.
Am I right?

**Comment by EECOLOR** [comment]
We do a lot of visual appealing websites with transitions all over the place. To make sure we do not enter hard to debug problems we have the following rules:

- DOM writes in `requestAnimationFrame` or `useLayoutEffect`
- DOM reads everywhere else

In some cases we need to set a style property on the DOM, for example `overflow: hidden` when we have a modal or full screen menu. Our effect would look like something this:

```js
const isMounted = useRenderIfMounted() // returns true if mounted
useLayoutEffect(
  () => {
    if (isMounted) {
      // perform writes
    }
    () => { /* undo writes */ }
  }.
  [isMounted]
)
```

As you can see there is a guard to make sure it does not run on the server.

**Comment by JoostKiens** [comment]
A quick solution is a component which renders another component after mounting. The second component executes the effect and returns `null`.

```jsx
function UseLayoutEffect({ effect, dependencies }) {
  const isMounted = useRenderIfMounted() // returns true if mounted
  return isMounted && <ActualEffect {...{ effect, dependencies }} />
}

function ActualEffect({ effect, dependencies }) {
  React.useLayoutEffect(effect, dependencies)
  return null
}
```

**Comment by EECOLOR** [comment]
The workaround is usable, but it nullifies the eslinting rules that help tremendously.

Since our default mode is server-side rendering and then adding universal stuff for the interactive parts, it would really help if could somehow disable the warning. I think the best solution would actually be a `React.useMountedLayoutEffect(...)`

**Comment by EECOLOR** [comment]
Just thinking out loud. I think it is easier to this:

```js
const isBrowser = typeof window !== 'undefined'
if (isBrowser) React.useLayoutEffect(...)
```

While the rules state a hook should not be used within an `if` statement, that rule is only in place to make sure any memory slots assigned do not get mixed up. This `if` statement does not cause problems because the `isBrowser` switch will never change within an environment.

React has the boolean `canUseDOM`, so theoretically we could use that. It would be ideal if the `if` statement was in `React.useLayoutEffect` itself. This however does not work if you take other forms of rendering into account (non-dom rendering).

If I was the library author I would explore the idea of asking the renderer (is that Fiber?) if it supports `useLayoutEffect`. The default renderer could then use the `canUseDOM` while the native renderer (or canvas renderer) could use another mechanism.

**Comment by eps1lon** [comment]
@EECOLOR 
It warns not depending on DOM detection but if a certain API is used (`react-dom/server`). You can run `ReactDOM.renderToStaticMarkup` in the browser which will trigger this warning. No environment sniffing can help you mitigate this.

**Comment by EECOLOR** [comment]
@eps1lon I understand what you are saying.

I don't think that rendering a component both on the server and on the client is uncommon. Using `useEffect` does not result in a warning because it will only be executed after the paint (which does not happen on the server).

It seems to be well known that `useEffect` has no effect when a component is rendered on the server.
This should be the same for `useLayoutEffect`, there is no way use cases for using it (writing to the DOM or other global browser API's) would succeed in a server environment.

I am hoping that `useLayoutEffect` can get to a similar state as `useEffect`: that it will not be executed when there is no need for it. At the moment this is up to the user of the library and the warning helps the user of the library to not forget fixing the problem. I think we can do better relatively easy, it would move the burden from the user (developer) to an if-statement.

**Comment by 4px** [comment]
I think the point of the warning is that the server rendering DOM and DOM built on the client will be different when the layoutEffect does something with DOM. But there are times when you need to do something before browser rendering. For example set scrollTo(0, 0) or set attributes of a title and meta tags (SSR does it too, but without effects). With useEffect there is a delay.
in this case:
```js
import React from 'react';

// Skip layoutEffects on server rendering to prevent warning
const isServerSide = typeof window == 'undefined';
const useLayoutEffect = isServerSide ? () => {} : React.useLayoutEffect;

...
```

**Comment by EECOLOR** [comment]
@4px I afree with what you are saying. In my mind it however does not make sense if you have to do this in every project while it can be solved by the library.

Another (maybe) less intrusive option would be to give the ability to disable the warning. This however is a bit more dangerous because people might forget to add the following header to the effect function:

```js
React.useLayoutEffect(
  () => {
    if (typeof window !== 'undefined') return
    ...
  },
  []
)
```

**Comment by markerikson** [comment]
React-Redux has actually just run into a potential bug, as our current feature-detection logic appears to mis-identify RN as Node:

https://github.com/reduxjs/react-redux/issues/1437

The current implementation is:

```js
const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' &&
  typeof window.document !== 'undefined' &&
  typeof window.document.createElement !== 'undefined'
    ? useLayoutEffect
    : useEffect
```

We may need to add a check for RN in there, possibly something like `navigator && navigator.product = 'ReactNative'` (per https://stackoverflow.com/a/39473604/62937 ).

Anyone else see something like this?

**Comment by eps1lon** [comment]
> Anyone else see something like this?

The problem generally is that this warning depends on what API you use not what environment you're in. So anything relying on environment sniffing will never be complete since you can use most of `react-dom/server` in the browser or in a node test environment with a global jsdom.

> And yes, all these shenanigans wouldn't be necessary if this warning just wasn't printed in the first place.

Agreed. The issue described in the warning cannot be fixed anyway. You get layout shift whether you use useLayoutEffect, or useEffect or csr only mounting in any way.  The workarounds in the docs do not fix the issue at all but make them more obvious. Apparently it's better to always have a layout shift instead of only once? 

Especially statements like

> This way, the UI doesn’t appear broken before hydration.

are incredibly frustrating. Just because we have to handle layout on the client doesn't mean the UI is "broken". And deferring it to a useEffect does not magically make the UI not broken. I guess this is where something like an enhanced effect hook would be nice: useEffect when hydrating but useLayoutEffect on render.

**Comment by aleclarson** [comment]
The easiest solution for library authors is to use [this package](https://www.npmjs.com/package/react-layout-effect).

**Comment by atomiks** [comment]
Can't believe that `useLayoutEffect` needs to be imported from a separate library to get things to work without issues. The warning should have been removed shortly ages ago after this issue was made, since it's clear the benefits it provides are minimal (or even misguided) compared to the problems it's caused for dozens of libs that can't work around it right now.

**Comment by necolas** [comment]
Closing this as answered.

Facebook essentially uses [this strategy](https://github.com/facebook/react/issues/14927#issuecomment-466815232) internally, and solving for the case where an effect condition can only happen due to an update is something we will look to support in the future as we introduce new APIs.

**Comment by gaearon** [comment]
To clarify, the link above is **an escape hatch** and should only be used when you understand the consequences. Before you use this, read this part of the docs:

>If you use server rendering, keep in mind that neither `useLayoutEffect` nor `useEffect` can run until the JavaScript is downloaded. This is why React warns when a server-rendered component contains `useLayoutEffect`. To fix this, either move that logic to useEffect (if it isn’t necessary for the first render), or delay showing that component until after the client renders (if the HTML looks broken until useLayoutEffect runs).

>To exclude a component that needs layout effects from the server-rendered HTML, render it conditionally with `showChild && <Child />` and defer showing it with `useEffect(() => { setShowChild(true); }, [])`. This way, the UI doesn’t appear broken before hydration.

If you absolutely need to work around it, sure, https://github.com/facebook/react/issues/14927#issuecomment-466815232 works. I'd say `useIsomorphicLayoutEffect` is an extremely confusing name for this pattern. At FB, this Hook is called `useLayoutEffect_SAFE_FOR_SSR` which calls out that you understand the consequences. Maybe you'll want to name it something similar.

**Comment by markerikson** [comment]
Is there a definitive implementation of the `isBrowser` check that can be officially recommended?

For React-Redux, we've had to keep making this more complicated, and run into issues due to folks running tests with Jest+JSDOM, as well as handling React Native.

Issues for reference:

- https://github.com/reduxjs/react-redux/issues/1402
- https://github.com/reduxjs/react-redux/issues/1492
- https://github.com/reduxjs/react-redux/issues/1437
- https://github.com/reduxjs/react-redux/pull/1444

Frankly, this warning continues to be a pain for us to deal with.

**Comment by StevieJayCee** [comment]
Giving this another bump, as `useIsomorphicLayoutEffect` gets confused by Jest+JSDOM as @markerikson pointed out. We either have the test logs polluted by this error warning or have a separate maintenance overhead with separated test classes for SSR, annotated with the alternative jest node environment.

**Comment by iDVB** [comment]
@gaearon Regarding the [tweak to the docs](https://github.com/facebook/react/issues/14927#issuecomment-474479615) you made. I've been reading up a fair bit on the useLayoutEffect warning and SSR conversation.  I think I fully understand the "what's happening" but still completely confused as to the why have a warning for **only useLayoutEffect**

From those docs...
> "keep in mind that neither useLayoutEffect **nor useEffect** can run until the JavaScript is downloaded. This is why React warns when a server-rendered component contains **useLayoutEffect**. "

Why then are we providing a warning only for useLayoutEffect? **Both** will cause renders on the client and not on the server. 

The most I can reason, is that the react team have identified that many people are using `useLayoutEffect` when it would be better to use `useEffect`. This seems like something better to leave to docs and education rather than forced warnings. 

For us, we explicitly use `useLayoutEffect` in order to setup complex timelined animations. 
Using useEffect causes blinking between routes and on mount, as would lazy loading those animations. 
Using useLayoutEffect has served us perfectly for years by rendering the animations, element locations etc before things render to the browser, so there is no blink.  So the solution for us to avoid these warnings is to implement [isomorphicLayoutEffect](https://github.com/streamich/react-use/blob/master/docs/useIsomorphicLayoutEffect.md) like switches. 

It seems there a plenty of valid use cases for `useLayoutEffect` to warrant finding a more subtle / off-build way to advise developers of when best to use it. No?

**Comment by just-boris** [comment]
Currently the warning does not work for the community

1. [`use-isomorphic-layout-effect` npm package](https://www.npmjs.com/package/use-isomorphic-layout-effect) exists for the sole purpose of hiding this warning and it has 2.6 M weekly downloads, 20% of total react downloads. So, every 5th project includes this dependency. 
2. There is also a whole lot of custom versions of this hack (https://github.com/facebook/react/issues/14927#issuecomment-466815232) ([github search](https://github.com/search?q=useIsomorphicLayoutEffect+window+undefined+extension%3Ajs+extension%3Ajsx+extension%3Ats+extension%3Atsx&type=Code))
3. [The most popular answer on stackoverflow](https://stackoverflow.com/a/58173551/1297743) also suggests to circumvent the warning by swapping the hook

I think, I understand the original intent of introducing this warning to prevent UI jumps. But apparently the community does not benefit from this warning, it is mostly seen as an obstacle.

Maybe it is time to revisit the warning, if its only purpose is to be suppressed by various workarounds?

**Comment by Kjaer** [comment]
> Currently the warning does not work for the community
> 
> 1. [`use-isomorphic-layout-effect` npm package](https://www.npmjs.com/package/use-isomorphic-layout-effect) exists for the sole purpose of hiding this warning and it has 2.6 M weekly downloads, 20% of total react downloads. So, every 5th project includes this dependency.
> 2. There is also a whole lot of custom versions of this hack ([useLayoutEffect in ssr #14927 (comment)](https://github.com/facebook/react/issues/14927#issuecomment-466815232)) ([github search](https://github.com/search?q=useIsomorphicLayoutEffect+window+undefined+extension%3Ajs+extension%3Ajsx+extension%3Ats+extension%3Atsx&type=Code))
> 3. [The most popular answer on stackoverflow](https://stackoverflow.com/a/58173551/1297743) also suggests to circumvent the warning by swapping the hook
> 
> I think, I understand the original intent of introducing this warning to prevent UI jumps. But apparently the community does not benefit from this warning, it is mostly seen as an obstacle.
> 
> Maybe it is time to revisit the warning, if its only purpose is to be suppressed by various workarounds?

I wouldn't rely on isomorphic layout effect package. It's not maintaining regularly and it switches between `useEffect` and `useLayoutEffect` based on the availability of `document`.
https://github.com/Andarist/use-isomorphic-layout-effect/blob/main/src/index.ts

so  Abramov's gist more valid than the isomorphic package.
https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
